ambry-store/src/main/java/com.github.ambry.store/HardDeleter.java

```diff
@@ -217,6 +221,7 @@ void performRecovery() throws StoreException {
         readOptionsList.add(item.blobReadOptions);
         messageStoreRecoveryInfoList.add(item.messagesStoreRecoveryInfo);
       }
+      hardDeleteRecoveryRange.clear();
```

Comment:
Could you briefly explain why this would fix NPE in `pruneTill()`? You can add the explanation to the commit message of this PR.

Reply:
updated.

Reply:
Got it. I wonder why `readCleanupTokenAndPopulateRecoveryRange` didn't pass in `startToken` when creating recoverRange previously? Looks like it's viable to do so. Nvm, will approve this PR.